src/qa-certification-workspace/qa-certification.service.ts

```diff
@@ -35,152 +51,296 @@ export class QACertificationWorkspaceService {
-      (qaTestExtensionExemptionId, idx) => {
-        promises.push(
-          new Promise((resolve, _reject) => {
-            const locationId = locations.find(i => {
-              return (
-                i.unitId === qaTestExtensionExemptionId.unitId &&
-                i.stackPipeId === qaTestExtensionExemptionId.stackPipeId
-              );
-            }).locationId;
-
-            const results = this.testExtensionExemptionService.import(
-              locationId,
-              qaTestExtensionExemptionId,
-              userId,
-            );
-            resolve(results);
-          }),
+    // Use transaction to ensure atomic operations
+    return this.entityManager.transaction(async (trx: EntityManager) => {
+      try {
+        this.logger.log(
+          `Starting QA Certification import transaction for Facility Id/Oris Code [${payload.orisCode}]`,
         );
-      },
-    );
-    payload.certificationEventData?.forEach((qaCertEvent, idx) => {
-      promises.push(
-        new Promise((resolve, _reject) => {
-          const locationId = locations.find(i => {
-            return (
-              i.unitId === qaCertEvent.unitId &&
-              i.stackPipeId === qaCertEvent.stackPipeId
-            );
-          }).locationId;
-
-          const results = this.qaCertEventService.import(
-            locationId,
-            qaCertEvent,
-            userId,
-          );
-
-          resolve(results);
-        }),
-      );
-    });
 
-    await Promise.all(promises);
+        // Process all data types in parallel and collect errors
+        const promises = [
+          this.processTestSummaryData(locations, payload.testSummaryData, userId, qaSupprecords, trx),
+          this.processTestExtensionData(locations, payload.testExtensionExemptionData, userId, trx),
+          this.processCertificationEventData(locations, payload.certificationEventData, userId, trx),
+        ];
 
-    return {
-      message: `Successfully Imported QA Certification Data for Facility Id/Oris Code [${payload.orisCode}]`,
-    };
+        // Use Promise.allSettled to collect all results and errors
+        const results = await Promise.allSettled(promises);
+        const errors: Error[] = [];
+
+        // Process the results
+        results.forEach((result, index) => {
+          if (result.status === 'rejected') {
+            const error = result.reason;
+            if (this.logger) {
+              this.logger.error(`Error in promise at index ${index}: ${error.message}`);
+            }
+            errors.push(error);
+          }
+        });
+
+        // If there were errors, log a summary
+        if (errors.length > 0 && this.logger) {
+          this.logger.error(`${errors.length} errors occurred while processing promises`);
+        }
+
+        // If rethrowErrors is true and there are errors, throw the first one
+        if (rethrowErrors && errors.length > 0) {
+          throw errors[0];
+        }
```

Comment:
This `rethrowErrors` flag should not be here: the errors should always be re-thrown so that that the transaction is aborted.
```suggestion
        if (errors.length > 0) {
          throw errors[0];
        }
```